src/logger.js

```diff
@@ -1,23 +1,45 @@
 import createDebug from 'debug'
 import { discoveryKey } from 'hypercore-crypto'
+import mapObject from 'map-obj'
+import util from 'util'
 
 const TRIM = 7
 
-createDebug.formatters.h = (v) => {
+createDebug.formatters.h = function (v) {
   if (!Buffer.isBuffer(v)) return '[undefined]'
   return v.toString('hex').slice(0, TRIM)
 }
 
-createDebug.formatters.S = (v) => {
+createDebug.formatters.S = function (v) {
   if (typeof v !== 'string') return '[undefined]'
   return v.slice(0, 7)
 }
 
-createDebug.formatters.k = (v) => {
+createDebug.formatters.k = function (v) {
   if (!Buffer.isBuffer(v)) return '[undefined]'
   return discoveryKey(v).toString('hex').slice(0, TRIM)
 }
 
+/**
+ * @param {import('./sync/sync-state.js').State} v
+ * @this {any} */
+createDebug.formatters.X = function (v) {
+  const mapped = mapObject(v, (k, v) => [
+    k,
+    mapObject(v, (k, v) => {
```

Comment:
I'm going to just ts-ignore this - the type checking doesn't get us anything here, since this is never type checked anyway (v could actually be `any`). We could do full validation of `unknown` to ensure that `v` is `State` and fix all the type issues, but that would really slow down the code (debugging state already causes a massive slowdown), so I think it's better to just ignore errors.